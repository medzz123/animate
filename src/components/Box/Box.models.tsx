import { ResponsiveProps } from '@theme/ResponsiveProps/responsiveProps.models';
import { tokens } from '@theme/tokens';
import { CSSProperties } from 'react';

export interface BoxPropsExtended
  extends React.HTMLAttributes<HTMLDivElement>,
    Partial<ResponsiveProps<BoxProps>> {}

type SizeProp = keyof typeof tokens.sizes;

export interface BoxProps {
  /**
   * Margin
   */
  margin: SizeProp;
  /**
   * Margin Top
   */
  marginTop: SizeProp;
  /**
   * Margin Bottom
   */
  marginBottom: SizeProp;
  /**
   * Margin Left
   */
  marginLeft: SizeProp;
  /**
   * Margin Right
   */
  marginRight: SizeProp;
  /**
   * Padding
   */
  padding: SizeProp;
  /**
   * Padding Top
   */
  paddingTop: SizeProp;
  /**
   * Padding Bottom
   */
  paddingBottom: SizeProp;
  /**
   * Padding Left
   */
  paddingLeft: SizeProp;
  /**
   * Padding Right
   */
  paddingRight: SizeProp;
  /**
   * Width
   */
  width: SizeProp;
  /**
   * Height
   */
  height: SizeProp;
  /**
   * Max Width
   */
  maxWidth: SizeProp;
  /**
   * Max Height
   */
  maxHeight: SizeProp;
  /**
   * Min Width
   */
  minWidth: SizeProp;
  /**
   * Min Height
   */
  minHeight: SizeProp;
  /**
   * Display
   * @see — https ://developer.mozilla.org/docs/Web/CSS/display
   */
  display: CSSProperties['display'];
  /**
   * Justify content
   * @see — https ://developer.mozilla.org/docs/Web/CSS/justify-content
   */
  justifyContent: CSSProperties['justifyContent'];
  /**
   * Align items
   * @see — https ://developer.mozilla.org/docs/Web/CSS/align-items
   */
  alignItems: CSSProperties['alignItems'];
  /**
   * Align content
   * @see — https ://developer.mozilla.org/docs/Web/CSS/align-content
   */
  alignContent: CSSProperties['alignContent'];
  /**
   * Flex wrap
   * @see — https ://developer.mozilla.org/docs/Web/CSS/align-wrap
   */
  flexWrap: CSSProperties['flexWrap'];
  /**
   * Flex direction
   * @see — https ://developer.mozilla.org/docs/Web/CSS/flex-direction
   */
  flexDirection: CSSProperties['flexDirection'];
  /**
   * Flex
   * @see — https ://developer.mozilla.org/docs/Web/CSS/flex
   */
  flex: CSSProperties['flex'];
  /**
   * Position
   * @see — https ://developer.mozilla.org/docs/Web/CSS/position
   */
  position: CSSProperties['position'];
  /**
   * Position Absolute - Top
   *
   */
  top: SizeProp;
  /**
   * Position Absolute - Right
   *
   */
  right: SizeProp;
  /**
   * Position Absolute - Bottom
   *
   */
  bottom: SizeProp;
  /**
   * Position Absolute - Left
   *
   */
  left: SizeProp;
}
