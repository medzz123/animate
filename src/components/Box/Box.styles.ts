import styled from 'styled-components';

export const BoxContainer = styled.div<{ parsedCss: string }>`
  ${(p) => p.parsedCss}
`;
