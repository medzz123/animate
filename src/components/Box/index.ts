import styled from 'styled-components';

interface BoxProps {
  m?: number;
  mt?: number;
  mb?: number;
  ml?: number;
  mr?: number;

  p?: number;
  pt?: number;
  pb?: number;
  pl?: number;
  pr?: number;

  w?: number;
  h?: number;
}

const Box = styled.div<BoxProps>`
  margin-left: ${(p) => p.ml}px;
  margin-right: ${(p) => p.mr}px;
  margin-bottom: ${(p) => p.mb}px;
  margin-top: ${(p) => p.mt}px;
  margin: ${(p) => p.m}px;

  padding-left: ${(p) => p.pl}px;
  padding-right: ${(p) => p.pr}px;
  padding-bottom: ${(p) => p.pb}px;
  padding-top: ${(p) => p.pt}px;
  padding: ${(p) => p.p}px;

  width: ${(p) => p.w}px;
  height: ${(p) => p.h}px;
`;

export default Box;
