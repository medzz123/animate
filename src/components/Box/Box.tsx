import { forwardRef, useMemo } from 'react';

import { replaceTokens } from '../../theme/ReplaceTokens';
import { responsiveProps } from '../../theme/ResponsiveProps';
import { BoxProps, BoxPropsExtended } from './Box.models';
import { BoxContainer } from './Box.styles';

const Box = forwardRef<HTMLDivElement, BoxPropsExtended>((props, ref) => {
  const {
    children,
    margin,
    padding,
    marginBottom,
    marginTop,
    marginLeft,
    marginRight,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
    width,
    maxHeight,
    maxWidth,
    minHeight,
    minWidth,
    height,
    display = 'block',
    flexDirection,
    flexWrap,
    justifyContent,
    alignContent,
    alignItems,
    position,
    top,
    right,
    left,
    bottom,
    flex,
    ...rest
  } = props;

  const parsedProps = useMemo(() => {
    const replacedObject = replaceTokens({
      margin,
      padding,
      marginBottom,
      marginTop,
      marginLeft,
      marginRight,
      paddingBottom,
      paddingLeft,
      paddingRight,
      paddingTop,
      width,
      maxHeight,
      maxWidth,
      minHeight,
      minWidth,
      height,
      top,
      right,
      left,
      bottom,
    });

    return responsiveProps<BoxProps>({
      ...replacedObject,
      display,
      flexDirection,
      flexWrap,
      justifyContent,
      alignContent,
      alignItems,
      position,
      flex,
    });
  }, [
    margin,
    padding,
    marginBottom,
    marginTop,
    marginLeft,
    marginRight,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
    width,
    maxHeight,
    maxWidth,
    minHeight,
    minWidth,
    height,
    display,
    flexDirection,
    flexWrap,
    justifyContent,
    alignContent,
    alignItems,
    position,
    top,
    right,
    left,
    bottom,
    flex,
  ]);

  return (
    <BoxContainer ref={ref} parsedCss={parsedProps} data-testid="box" {...rest}>
      {children}
    </BoxContainer>
  );
});

export default Box;
