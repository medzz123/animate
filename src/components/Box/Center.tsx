import { forwardRef } from 'react';

import Box from './Box';
import { BoxPropsExtended } from './Box.models';

const Flex = forwardRef<HTMLDivElement, BoxPropsExtended>((props, ref) => (
  <Box
    display="flex"
    width="full"
    height="full"
    justifyContent="center"
    alignItems="center"
    ref={ref}
    {...props}
  >
    {props?.children}
  </Box>
));

Flex.displayName = 'Box-Center';

export default Flex;
