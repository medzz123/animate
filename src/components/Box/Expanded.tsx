import { forwardRef } from 'react';

import Box from './Box';
import { BoxPropsExtended } from './Box.models';

const Expanded = forwardRef<HTMLDivElement, BoxPropsExtended>((props, ref) => (
  <Box width="full" height="full" ref={ref} {...props}>
    {props?.children}
  </Box>
));

Expanded.displayName = 'Box-Expanded';

export default Expanded;
