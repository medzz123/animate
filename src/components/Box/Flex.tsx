import { forwardRef } from 'react';

import Box from './Box';
import { BoxPropsExtended } from './Box.models';

const Flex = forwardRef<HTMLDivElement, BoxPropsExtended>((props, ref) => (
  <Box display="flex" ref={ref} {...props}>
    {props?.children}
  </Box>
));

Flex.displayName = 'Box-Flex';

export default Flex;
