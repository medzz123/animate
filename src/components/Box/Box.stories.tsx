import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';

import Box from './Box';
import Expanded from './Expanded';

export default {
  title: 'Components/Box',
  component: Box,
} as Meta;

const Template: Story = (storyArguments) => (
  <Box {...storyArguments}>
    <p>of-</p>
    <p>course-</p>
    <p>i-</p>
    <p>still-</p>
    <p>love-</p>
    <p>you</p>
  </Box>
);

export const BoxStory = Template.bind({});

BoxStory.args = {
  display: 'flex',
  flexDirection: { 375: 'column', 480: 'row' },
  padding: { 480: 20, 992: 40 },
};

const ExpandedTemplate: Story = (storyArguments) => (
  <div style={{ width: 300, height: 300, backgroundColor: 'tomato' }}>
    <Expanded
      display="flex"
      justifyContent="center"
      alignItems="center"
      {...storyArguments}
    >
      <p>Hello</p>
    </Expanded>
  </div>
);

export const ExpandedStory = ExpandedTemplate.bind({});
