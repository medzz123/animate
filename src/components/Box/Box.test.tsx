import { render } from '@test/testUtils';
import React from 'react';

import Box from './Box';

test('Box Test', () => {
  const { getByTestId, container } = render(<Box>Hello</Box>);

  const boxNode = getByTestId('box');

  expect(boxNode.textContent).toBe('Hello');

  expect(container).toMatchSnapshot();
});
