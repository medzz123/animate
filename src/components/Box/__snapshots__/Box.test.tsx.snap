// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box Test 1`] = `
<div>
  <div
    class="sc-bdfBwQ JrvQt"
    data-testid="box"
  >
    Hello
  </div>
</div>
`;
